server/internal/worker.go

```diff
@@ -74,36 +78,107 @@ func getHttpServerPort() int32 {
+		line := fmt.Sprintf("[%s] %s", pw.prefix, scanner.Text())
+		// Write the prefixed line to the underlying writer
+		n, err := pw.writer.Write([]byte(line + "\n"))
+		totalWritten += n
+
+		if err != nil {
+			return totalWritten, err
+		}
+	}
+
+	// Check if the scanner encountered any error
+	if err := scanner.Err(); err != nil {
+		return totalWritten, err
+	}
+
+	return len(p), nil
+}
+
 func (w *Worker) start(req *StartReq) (err error) {
-	shell := fmt.Sprintf("cd /app/agents && nohup %s --property %s > %s 2>&1 &", workerExec, w.PropertyJsonFile, w.LogFile)
+	shell := fmt.Sprintf("cd /app/agents && %s --property %s", workerExec, w.PropertyJsonFile)
 	slog.Info("Worker start", "requestId", req.RequestId, "shell", shell, logTag)
-	if _, err = exec.Command("sh", "-c", shell).CombinedOutput(); err != nil {
-		slog.Error("Worker start failed", "err", err, "requestId", req.RequestId, logTag)
-		return
+	cmd := exec.Command("sh", "-c", shell)
+
+	var stdoutWriter, stderrWriter io.Writer
+	var logFile *os.File
+
+	if w.Log2Stdout {
+		// Write logs to stdout and stderr
+		stdoutWriter = os.Stdout
+		stderrWriter = os.Stderr
+	} else {
+		// Open the log file for writing
+		logFile, err := os.OpenFile(w.LogFile, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
+		if err != nil {
+			slog.Error("Failed to open log file", "err", err, "requestId", req.RequestId, logTag)
+			// return err
+		}
+
+		// Write logs to the log file
+		stdoutWriter = logFile
+		stderrWriter = logFile
 	}
 
-	shell = fmt.Sprintf("ps aux | grep %s | grep -v grep | awk '{print $2}'", w.PropertyJsonFile)
-	slog.Info("Worker get pid", "requestId", req.RequestId, "shell", shell, logTag)
+	// Create PrefixWriter instances with appropriate writers
+	stdoutPrefixWriter := &PrefixWriter{
+		prefix: "0", // Initial prefix, will update after process starts
+		writer: stdoutWriter,
+	}
+	stderrPrefixWriter := &PrefixWriter{
+		prefix: "0", // Initial prefix, will update after process starts
+		writer: stderrWriter,
+	}
 
-	var pid int
-	for i := 0; i < 3; i++ { // retry for 3 times
-		output, err := exec.Command("sh", "-c", shell).CombinedOutput()
-		if err == nil {
-			pid, err = strconv.Atoi(strings.TrimSpace(string(output)))
-			if err == nil && pid > 0 {
-				break // if pid is successfully obtained, exit loop
-			}
-		}
-		slog.Warn("Worker get pid failed, retrying...", "attempt", i+1, "requestId", req.RequestId, logTag)
-		time.Sleep(500 * time.Millisecond) // wait for 500ms
+	cmd.Stdout = stdoutPrefixWriter
+	cmd.Stderr = stderrPrefixWriter
+
+	if err = cmd.Start(); err != nil {
+		slog.Error("Worker start failed", "err", err, "requestId", req.RequestId, logTag)
+		return
 	}
 
+	pid := cmd.Process.Pid
+
 	if pid <= 0 {
 		slog.Error("Worker failed to obtain valid PID after 3 attempts", "requestId", req.RequestId, logTag)
 		return fmt.Errorf("failed to obtain valid PID")
 	}
 
+	// Update the prefix with the actual PID
+	stdoutPrefixWriter.prefix = fmt.Sprintf("%d", cmd.Process.Pid)
```

Comment:
Is it safe to modify the prefix since cmd has started?

Reply:
yes. a prefix writer is just a proxy for stdout writing, it should be ok to change it.

Reply:
But the `Write` will be called in another Goroutine, right? Personally I'll prefer to use the property json file name as prefix here since it's static and meaningful, even though the downside is too long.

Reply:
@wangyoucao577 yes. the property json file is unnecessarily long and meaningless. i can put channel name as prefix? maybe that's easier to understand. the purpose is to distinguish logs from different processes.

Reply:
@plutoless  OK, it looks like a reasonable trade-off, let's move forward with it.